Create HOC to avoid code duplication in components

tonnguyen opened the pull request:
Applying 'editable' enhancer would transform a pure component to a component that can react in OPE mode.
Editable enhancer can wrap to other components too.

By introducing HOC, we reduce the complexity of the component, make it pure, and easier to test.

seriema commented on AlloyReact/Static/react/editable.js:
```diff
@@ -0,0 +1,60 @@
+    }
+}
+
+const editable = (WrappedComponent, onValueChange = defaultValueUpdator) => {
```
I like the idea! Should we update the other examples (`ThreeKeyFacts.jsx` and `PageListBlock.jsx`) to use this as well?

tonnguyen replied:
Yes we definitely should

seriema commented on AlloyReact/Helpers/HtmlHelpers.cs:
```diff
@@ -25,19 +25,20 @@ public static class HtmlHelpers
         /// <remarks>This HTML helper also ensure required script resources are loaded</remarks>
         /// <returns></returns>
-        public static IHtmlString ReactComponent(this HtmlHelper html, string component, object props = null)
+        public static IHtmlString ReactComponent(this HtmlHelper html, string propertyName, string component, object props = null)
```
Maybe you can just use the existing `ReactComponentFor` instead? :)

tonnguyen replied:
But then there should be no different between the output HTML in view or edit mode.

seriema replied:
Why do you want the `data-epi-*` attributes to show up on the public site?

tonnguyen replied:
We need it in order to map a page propery to a component’s prop. We have a page propery named someNumber, and it can be anything that we don’t now at compile time. In the component, we have a prop named number. By returning the page property, we know what propery value should we get and map to the component’s prop.

seriema replied:
I'm sorry, I still don't understand how that is affected by not being in edit mode? I'd assume that a partner would not want the `data-epi-*` properties to be shown when viewing the site. To only want to use our OPE editing functionality when in OPE?

tonnguyen replied:
Take a look at ReactNumber.jsx, this.props.propertyName is empty in View mode. And it was hard-coding the property name in props: props.someNumber.

'someNumber' is a property name, defined in the StandardPage model. I don't think we can use that fixed string in the component.

ben-mckernan replied:
Yes, but the editable HOC uses `props.propertyName` in order to retrieve data and set it back to the given `propName` on the component. In this case "number". You don't read the `data-epi-property-name` anywhere that I can see. So why would it be needed outside of edit mode?

tonnguyen replied:
No I don't have to read the attribute, but props.propertyName will be null then. Without my changes, props.propertyName is available in Edit mode only.

seriema commented on AlloyReact/Views/Shared/Blocks/ThreeKeyFactsBlock.cshtml:
```diff
@@ -2,6 +2,7 @@
 @model ThreeKeyFactsBlock
 
 @Html.ReactComponent(
```
Here you can use `@Html.ReactComponentFor(` instead.

seriema commented on AlloyReact/Static/react/ReactNumber.jsx:
```diff
@@ -1,44 +1,11 @@
-export default class ReactBlock extends React.Component {
+const ReactBlock = ({ propertyName, number }) => 
+    <div>
+        The current value of the <strong>{propertyName}</strong> property is <strong>{number}</strong>, and the current time is: <strong><Clock /></strong>
```
Minor thing: can you keep the first sentence as well? It's there to show the difference in timestamps between first render and future updates, to show how the component is updated more efficiently with the contentSaved event.

`This React component was rendered at <strong>{this.state.firstRender.toLocaleTimeString("sv-SE")}</strong> with a number of <strong>{this.props.someNumber}</strong>.<br/><br/>`

seriema commented on AlloyReact/Static/react/editable.js:
```diff
@@ -0,0 +1,60 @@
+import hoistStatics from 'hoist-non-react-statics';
+import React, { Component } from 'react';
+
+const defaultValueUpdator = (value, propName, props, propertyDetails = null) => {
```
Minor spelling: `defaultValueUpdater`

ton-litium committed:
Create HOC to avoid code duplication in components

Applying 'editable' enhancer would turn a pure component to a component
that can react in OPE mode.
Editable enhancer can wrap to other components too.

seriema commented:
This PR has gone stale. Closing. The CMS UI team will also focus more on the MusicFestival template site instead. Please help there if you have time. https://github.com/episerver/musicfestival-vue-template
